xpallets/gateway/bitcoin/v2/src/vault.rs

```diff
@@ -94,75 +110,79 @@ pub mod vault {
                 collateral >= Self::minimium_vault_collateral(),
                 Error::<T>::InsufficientVaultCollateralAmount
             );
+            ensure!(!Self::vault_exists(&sender), Error::<T>::VaultRegistered);
+            ensure!(
+                !Self::btc_address_exists(&btc_address),
+                Error::<T>::BtcAddressOccupied
+            );
             Self::lock_collateral(&sender, collateral)?;
+            Self::increase_total_collateral(collateral);
+            Self::insert_btc_address(&btc_address, sender.clone());
             let vault = Vault::new(sender.clone(), btc_address);
             Self::insert_vault(&sender, vault.clone());
-            Self::deposit_event(vault.id, collateral);
-            //TODO(wangyafei)
+            Self::deposit_event(Event::RegisterVault(vault.id, collateral));
             Ok(().into())
         }
     }
 
     #[pallet::error]
     pub enum Error<T> {
+        /// Requester doesn't has enough pcx for collateral.
         InsufficientFunds,
+        /// The amount in request is less than minimium bound.
         InsufficientVaultCollateralAmount,
+        /// Requester has been vault.
+        VaultRegistered,
+        /// Btc address in request was occupied by another vault.
+        BtcAddressOccupied,
     }
 
     #[pallet::event]
-    // Additional argument to specify the metadata to use for given type.
     #[pallet::metadata(BalanceOf<T> = "Balance", u32 = "Other")]
-    // Generate a funciton on Pallet to deposit an event.
     #[pallet::generate_deposit(pub(super) fn deposit_event)]
     pub enum Event<T: Config> {
         RegisterVault(<T as frame_system::Config>::AccountId, PCX<T>),
     }
 
     #[pallet::type_value]
-    pub(super) fn zero_pcx<T: Config>() -> PCX<T> {
+    pub(super) fn ZeroPcx<T: Config>() -> PCX<T> {
         0.into()
```

Comment:
```suggestion
        0u32.into()
```

With `u32` it can not compile with the latest Rust.